### PR TITLE
[JENKINS-15][JENKINS-28] Added option to notify repeated failure

### PR DESCRIPTION
**Pull request by peergum**
Option is set to true by default.
Also added two test connection buttons in global and build slack settings.
And finally, did some code cleanup at places.

### CODE REVIEW

File: src/main/java/jenkins/plugins/slack/ActiveNotifier.java [hunk 1]
```diff
@@ -60,9 +60,9 @@ private void notifyStart(AbstractBuild build, String message) {
         AbstractProject<?, ?> project = build.getProject();
         AbstractBuild<?, ?> previousBuild = project.getLastBuild().getPreviousBuild();
         if (previousBuild == null) {
-           getSlack(build).publish(message, "good");
+            getSlack(build).publish(message, "good");
```
Comment (samrocketman) [inline src/main/java/jenkins/plugins/slack/ActiveNotifier.java]: Good indenting catch :)

File: src/main/java/jenkins/plugins/slack/ActiveNotifier.java [hunk 2]
```diff
@@ -77,14 +77,19 @@ public void completed(AbstractBuild r) {
...
+        } while (previousBuild != null && previousBuild.getResult() == Result.ABORTED);
         Result previousResult = (previousBuild != null) ? previousBuild.getResult() : Result.SUCCESS;
         if ((result == Result.ABORTED && jobProperty.getNotifyAborted())
-                || (result == Result.FAILURE && jobProperty.getNotifyFailure())
+                || (result == Result.FAILURE
+                && (previousResult != Result.FAILURE || jobProperty.getNotifyRepeatedFailure())
+                && jobProperty.getNotifyFailure())
                 || (result == Result.NOT_BUILT && jobProperty.getNotifyNotBuilt())
                 || (result == Result.SUCCESS
-                		&& (previousResult == Result.FAILURE || previousResult == Result.UNSTABLE)
```
Comment (samrocketman) [inline src/main/java/jenkins/plugins/slack/ActiveNotifier.java]: Death to the mix of tabs and spaces :D.  In `vim`, I use `:%s/\t/    /g` to replace them when I remember to.

File: src/main/java/jenkins/plugins/slack/ActiveNotifier.java [hunk 3]
```diff
@@ -171,13 +176,14 @@ String getBuildStatusMessage(AbstractBuild r, boolean includeTestSummary) {
         message.appendStatusMessage();
         message.appendDuration();
         message.appendOpenLink();
-        if (!includeTestSummary){
+        if (!includeTestSummary) {
```
Comment (samrocketman) [inline src/main/java/jenkins/plugins/slack/ActiveNotifier.java]: Consistent code bracket spacing is great.

File: src/main/java/jenkins/plugins/slack/ActiveNotifier.java [hunk 4]
```diff
@@ -201,12 +207,24 @@ static String getStatusMessage(AbstractBuild r) {
             Result result = r.getResult();
             Run previousBuild = r.getProject().getLastBuild().getPreviousBuild();
             Result previousResult = (previousBuild != null) ? previousBuild.getResult() : Result.SUCCESS;
-            if (result == Result.SUCCESS && previousResult == Result.FAILURE) return "Back to normal";
-            if (result == Result.SUCCESS) return "Success";
-            if (result == Result.FAILURE) return "Failure";
-            if (result == Result.ABORTED) return "Aborted";
-            if (result == Result.NOT_BUILT) return "Not built";
-            if (result == Result.UNSTABLE) return "Unstable";
+            if (result == Result.SUCCESS && previousResult == Result.FAILURE) {
```
Comment (samrocketman) [inline src/main/java/jenkins/plugins/slack/ActiveNotifier.java]: Good code style.  Thanks for this!  Better to be explicit about conditionals.

File: src/main/java/jenkins/plugins/slack/SlackNotifier.java [hunk 5]
```diff
@@ -73,13 +76,16 @@ public BuildStepMonitor getRequiredMonitorService() {
...
 
     public SlackService newSlackService(String teamDomain, String token, String projectRoom) {
-    	// Settings are passed here from the job, if they are null, use global settings
-    	if (teamDomain == null)
-    		teamDomain = getTeamDomain();
-    	if (token == null)
-    		token = getAuthToken();
-    	if (projectRoom == null)
-    		projectRoom = getRoom();
+        // Settings are passed here from the job, if they are null, use global settings
```
Comment (samrocketman) [inline src/main/java/jenkins/plugins/slack/SlackNotifier.java]: Again great code style changes.  You're on a roll.

File: src/main/webapp/help-projectConfig-testConnection.html [hunk 6]
```diff
@@ -0,0 +1,4 @@
+<div>
```
Comment (samrocketman) [inline src/main/webapp/help-projectConfig-testConnection.html]: Thanks a lot for adding documentation to the options!

**Comment by samrocketman** [comment]
Candidate for merge means I will first merge your change locally, compile locally, and test locally in Jenkins before merging.

**Comment by samrocketman** [comment]
This will be in for the next release if all goes well in my testing.

File: src/main/java/jenkins/plugins/slack/ActiveNotifier.java (hunk 1 above)
Comment (peergum) [inline src/main/java/jenkins/plugins/slack/ActiveNotifier.java]: ctrl-shift-F in netbeans. It does it better than I would ;-)

File: src/main/java/jenkins/plugins/slack/ActiveNotifier.java (hunk 2 above)
Comment (peergum) [inline src/main/java/jenkins/plugins/slack/ActiveNotifier.java]: you'd rather use a s/\t/     /g (4 spaces) it looks nicer... didn't know about % I thought it was just the filename as in !cat % ... always used :1,$

**Comment by peergum** [comment]
Great. As I said, the code styling is all Netbeans. My part was only adding bugs in ;-)

**Comment by peergum** [comment]
For the note, I wish I had made the Repeat Failure Notification only showing when the Failure Notification is enabled (possibly below it, indented) but I'm not so familiar with Jelly... If anyone wants to make it better, feel free to patch accordingly.

**Comment by samrocketman** [comment]
I'm delaying the release until tomorrow.  Im currently not able to release right now.  Just giving you a heads up.

**Comment by samrocketman** [comment]
Currently working on the release.  One thing I see when I run this plugin is a bunch of "started" messages when I have ignore consecutive errors checked.  I would expect there to be no output.  Not even the started message.  Just my 2c.

jenkins_slack1

**Comment by samrocketman** [comment]
When you say JENKINS-15 and JENKINS-28 are you referring to the Jenkins issue tracker or the Slack GitHub project issues?  If you're referring to the GitHub project then you should simply do #15 and #28.

**Comment by peergum** [comment]
To the issue tracker, as information in Jenkins documentation ;-)

**Comment by peergum** [comment]
Referring your comment above, the check box is *Notify Repeated Failures* (which is the previous default), not *Ignore Repeated Failures*. I didn't want to go away from the current naming convention "notifyXXX" by using "doNotNotifyXXX"... Not sure it makes sense... :-)

**Comment by samrocketman** [comment]
Understood, I explained the feature for its function.  Perhaps that was a mistake.

The Jenkins documentation states when referencing issues at https://issues.jenkins-ci.org/.  In your case, you're not referencing issues at jenkins-ci.org.  You're referencing GitHub issues.  Please use `#[0-9]+` for GitHub issues and `JENKINS-[0-9]+` for jenkins-ci.org issues.

**Comment by peergum** [comment]
My bad. I misunderstood the jenkins bot was closing the issues here automatically...

**Comment by samrocketman** [comment]
No worries. If you include `fixes #[0-9]+` or `closes #[0-9]+` then GitHub will automatically close issues when pull requests are merged.

**Pull request merged**